File: packages/cli/hieudoanm/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/English.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.required
import com.google.gson.GsonBuilder
import io.github.hieudoanm.cli.services.Requests

class EnglishCommand : CliktCommand(name = "english", help = "English dictionary tools") {
    init {
        subcommands(EnglishDefine())
    }
    override fun run() = Unit
}

data class EnglishResult(
    val definition: String = "",
    val partOfSpeech: String = "",
    val synonyms: List<String> = emptyList(),
    val anonyms: List<String> = emptyList(),
    val usageOf: List<String> = emptyList(),
    val typeOf: List<String> = emptyList()
)

data class EnglishWord(
    val word: String = "",
    val results: List<EnglishResult> = emptyList()
)

class EnglishDefine : CliktCommand(name = "define", help = "Look up the definition of an English word") {
    private val word: String by option("--word", "-w", help = "Word to define").required()
    private val jsonOutput: Boolean by option("--json", help = "Output in JSON format").flag()

    override fun run() {
        val w = word.trim().lowercase()
        if (w.isEmpty()) {
            echo("word cannot be empty", err = true)
            return
        }
        val url = "https://raw.githubusercontent.com/hieudoanm/words/refs/heads/master/packages/data/english/words/$w.json"
        val result = Requests.get(url)
        if (result.isFailure) {
            echo("fetch error: ${result.exceptionOrNull()?.message}", err = true)
            return
        }
        val gson = GsonBuilder().setPrettyPrinting().create()
        val data = try {
            gson.fromJson(result.getOrNull(), EnglishWord::class.java)
        } catch (e: Exception) {
            echo("json error: ${e.message}", err = true)
            return
        }

        if (jsonOutput) {
            echo(gson.toJson(data))
            return
        }

        echo("\nWORD: ${data.word}\n")
        data.results.forEachIndexed { i, r ->
            echo("${i + 1}) ${r.partOfSpeech}")
            echo("   Definition: ${r.definition}")
            if (r.synonyms.isNotEmpty()) {
                echo("   Synonyms: ${r.synonyms.joinToString(", ")}")
            }
            if (r.anonyms.isNotEmpty()) {
                echo("   Antonyms: ${r.anonyms.joinToString(", ")}")
            }
            echo()
        }
    }
}

File: packages/cli/hieudoanm/cobra.go/cmd/english/english.go
package english

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"

	"github.com/hieudoanm/hieudoanm/libs/requests"
	"github.com/spf13/cobra"
)

type Result struct {
	Definition   string   `json:"definition"`
	PartOfSpeech string   `json:"partOfSpeech"`
	Synonyms     []string `json:"synonyms"`
	Anonyms      []string `json:"anonyms"`
	UsageOf      []string `json:"usageOf"`
	TypeOf       []string `json:"typeOf"`
}

type Word struct {
	Word    string   `json:"word"`
	Results []Result `json:"results"`
}

var defineCmd = &cobra.Command{
	Use:   "define <word>",
	Short: "Look up the definition of an English word",
	Long:  `Fetches and displays the definition, part of speech, synonyms, and antonyms for a given English word from a local dictionary data source.`,
	Args:  cobra.ExactArgs(1),
	RunE: func(cmd *cobra.Command, args []string) error {
		word := strings.TrimSpace(strings.ToLower(args[0]))
		if word == "" {
			return errors.New("word cannot be empty")
		}

		url := fmt.Sprintf(
			"https://raw.githubusercontent.com/hieudoanm/words/refs/heads/master/packages/data/english/words/%s.json",
			word,
		)

		body, err := requests.Get(url, requests.Options{})
		if err != nil {
			return fmt.Errorf("fetch error: %w", err)
		}

		var data Word
		if err := json.Unmarshal(body, &data); err != nil {
			return fmt.Errorf("json error: %w", err)
		}

		fmt.Printf("\nWORD: %s\n\n", data.Word)
		for i, r := range data.Results {
			fmt.Printf("%d) %s\n", i+1, r.PartOfSpeech)
			fmt.Printf("   Definition: %s\n", r.Definition)
			if len(r.Synonyms) > 0 {
				fmt.Printf("   Synonyms: %s\n", strings.Join(r.Synonyms, ", "))
			}
			if len(r.Anonyms) > 0 {
				fmt.Printf("   Antonyms: %s\n", strings.Join(r.Anonyms, ", "))
			}
			fmt.Println()
		}
		return nil
	},
}

func NewCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "english",
		Short: "English dictionary tools",
		Long:  `English dictionary lookup tool that fetches word definitions, synonyms, antonyms, and usage examples.`,
	}
	cmd.AddCommand(defineCmd)
	return cmd
}
